libraries/abstractions/secure_sockets/utest/secure_sockets_utest.c

```diff
@@ -0,0 +1,1334 @@
+{
+    uint32_t ret;
+
+    vPortFree_Stub( free_cb );
+    lwip_close_IgnoreAndReturn( 0 );
+    ret = SOCKETS_Close( so );
+    TEST_ASSERT_EQUAL_INT_MESSAGE( SOCKETS_ERROR_NONE, ret,
+                                   "ERROR: Could not close socket" );
+}
+
+/* helper function to initialized commonly used callbacks */
+static void initCallbacks( void )
+{
+    pvPortMalloc_Stub( malloc_cb );
+    vPortFree_Stub( free_cb );
+}
+
+/* helper function to uninitialize the commonly used callbacks */
+static void uninitCallbacks( void )
+{
+    pvPortMalloc_Stub( NULL );
+    vPortFree_Stub( NULL );
+}
+
+/* enforce tls, this is a one way option */
+static void setTLSMode( Socket_t s )
+{
+    int32_t ret = SOCKETS_SetSockOpt( s, 0, SOCKETS_SO_REQUIRE_TLS, NULL, 0 );
+
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static void setALPN_protocols( Socket_t s,
+                               int count )
+{
+    TEST_ASSERT_TRUE( count <= 2 );
+    char * pcAlpns[] = { socketsAWS_IOT_ALPN_MQTT, socketsAWS_IOT_ALPN_MQTT };
+
+    int32_t ret = SOCKETS_SetSockOpt( s,
+                                      0, /* Level - Unused. */
+                                      SOCKETS_SO_ALPN_PROTOCOLS,
+                                      pcAlpns,
+                                      count );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static void setServerNameIndication( Socket_t s )
+{
+    int32_t ret = SOCKETS_SetSockOpt( s,
+                                      0,
+                                      SOCKETS_SO_SERVER_NAME_INDICATION,
+                                      "destination",
+                                      11 );
+
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static void setServerCert( Socket_t s )
+{
+    int32_t ret = SOCKETS_SetSockOpt( s,
+                                      0,
+                                      SOCKETS_SO_TRUSTED_SERVER_CERTIFICATE,
+                                      "certificate",
+                                      11 );
+
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static void connectSocket( Socket_t s )
+{
+    int32_t ret;
+    SocketsSockaddr_t pxAddress;
+
+    lwip_connect_ExpectAnyArgsAndReturn( 0 );
+    ret = SOCKETS_Connect( s, &pxAddress, sizeof( SocketsSockaddr_t ) );
+    TEST_ASSERT_EQUAL_INT( SOCKETS_ERROR_NONE, ret );
+}
+
+static Socket_t create_TLS_connection()
+{
+    Socket_t s = initSocket();
+
+    setTLSMode( s );
+    TLS_Init_IgnoreAndReturn( pdFREERTOS_ERRNO_NONE );
+    TLS_Connect_IgnoreAndReturn( pdFREERTOS_ERRNO_NONE );
+    connectSocket( s );
+    return s;
+}
+
+static Socket_t create_normal_connection()
+{
+    Socket_t s = initSocket();
+
+    connectSocket( s );
+    return s;
+}
+/* ======================  TESTING SOCKETS_Send  ============================ */
+
+/*!
+ * @brief A happy send case with normal sockets
+ */
+void test01_SecureSockets_send_successful( void )
+{
```

Comment:
I want to make sure we establish patterns here that in the long run will be stable as these will become examples of how to do it. Do you think it is wise to have a test number as part of the test name like this? Keeping these numbers sequential and tracable seems to be hard e.g. when I delete a test in the middle somewhere, does that require me to rename all the tests? If not will it not look like there is a missing test?

Reply:
it was easier for me to find failed tests, but super annoying to renumber them
i removed the numbers..